### ISSUE TITLE
various artists logic

### ISSUE THREAD

**Issue by ofsaleem** [rdswift/picard-scripts#6]
hi, sorry to bombard you with so many PRs and issues this morning. first of all, thanks so much for your script, i've been using it with minimal edits for several years and it's straight up the only way i'd ever keep my library organized.

my issue pertains to, i think, this section of the script

```
$if($eq(%_nAlbumType%,Standard),$if($eq($if2(%musicbrainz_albumartistid%,%_cVariousArtistID%),%_cVariousArtistID%),
    $set(_nFilePath,%_cVariousArtist%/%_nYear% %_nANT%/),
    $if($eq($if2(%musicbrainz_albumartistid%,%_cUnknownArtistID%),%_cUnknownArtistID%),
        $set(_nFilePath,%_cUnknownArtist%/%_nYear% %_nANT%/),
        $set(_nFilePath,%_nInitial%/$if(%_aNoArtistSort%,%_nFAA%,%_nFAAPS%)/%_nYear% %_nANT%/)
)))
```

my understanding of this logic is:

check if the musicbrainz_albumartistid is available, if not grab the various artists ID constant
if neither the album artist id or the various artists ID constant are empty, set the path to use [Various Artists]
otherwise similar logic for setting the path to use [Unknown Artist] or the file path initials depending on user settings

but if i understand correctly, there are two issues here that are *i think* causing some problems with my files.

the first is that since `cVariousArtistID` is never empty since it is set in the file (i guess this check may be for robustness?), a file without the `musicbrainz_albumartistid` tag will *always* be set to Various Artists. afaiu the unknown artist logic never gets hit

the second is that i don't believe i follow the logic for treating the release as various artists because the artist ID isn't filled. i guess it is written this way as a failsafe in case MB doesn't have the album, because if it did, the tags would already be filled, but i think maybe it would be better to put this https://github.com/rdswift/picard-scripts/blob/d1ee2b13c932f47cc28595c6608db75cafe0bb84/naming_scripts/Bob_Swift/Bob%20Swift's%20Naming%20Script.pts#L241 here where you check if a track has a different artist than the album (this is a great feature btw), ie, if any one track meets this criteria then you could conceivably argue that the album should be categorized as Various Artists?

either way in my experience albums that aren't on MB get sorted as Various Artists, and in tandem with the soundtracks issue i described in https://github.com/rdswift/picard-scripts/pull/5 , my non-MB soundtracks will get thrown into Various Artists and will have the solo artist of the album listed in the filename in [Brackets] as if they were different from the main album artist

**Comment by rdswift** [comment]
I think that you may not be completely understanding the logic (or else I don't understand your explanation).  The first line uses the `$if2()` to get the `%musicbrainz_albumartistid%` if it exists, otherwise `%_cVariousArtistID%`, and then compares this value to `%_cVariousArtistID%`.  If it matches, the second line `$set(_nFilePath,%_cVariousArtist%/%_nYear% %_nANT%/)` is executed.  If it doesn't match (i.e. not Various Artists), then a similar check is done comparing `%musicbrainz_albumartistid%` (if it exists, otherwise `%_cUnknownArtistID%`) to `%_cUnknownArtistID%`.  If it matches the line 4 `$set(_nFilePath,%_cUnknownArtist%/%_nYear% %_nANT%/)` is executed.  If it doesn't match, then line 5 `$set(_nFilePath,%_nInitial%/$if(%_aNoArtistSort%,%_nFAA%,%_nFAAPS%)/%_nYear% %_nANT%/)` is executed.  In short:

| Album Artist | Format Used |
| --- | --- |
| Various Artists or blank | $set(_nFilePath,%_cVariousArtist%/%_nYear% %_nANT%/) |
| Unknown Artist | $set(_nFilePath,%_cUnknownArtist%/%_nYear% %_nANT%/) |
| Other | $set(_nFilePath,%_nInitial%/$if(%_aNoArtistSort%,%_nFAA%,%_nFAAPS%)/%_nYear% %_nANT%/) |

> the first is that since cVariousArtistID is never empty since it is set in the file (i guess this check may be for robustness?

Not exactly.  I think you're perhaps not understanding the `$if2()` statement.  That function returns the first non-empty item in the list provided to the function.  That way, if `%musicbrainz_albumartistid%` is empty, it returns `%_cVariousArtist%` instead (as the default value).

>  a file without the musicbrainz_albumartistid tag will always be set to Various Artists.

Correct.

> afaiu the unknown artist logic never gets hit

Not so.  If the `%musicbrainz_albumartistid%` variable is not empty and it is not equal to `%_cVariousArtist%` then the `%_cUnknownArtist%` check is performed.  I do see that the `$if2()` function is redundent in this second check because `%musicbrainz_albumartistid%` should never be blank at that point.  The line should instead be `$if($eq(%musicbrainz_albumartistid%,%_cUnknownArtistID%),`

> the second is that i don't believe i follow the logic for treating the release as various artists because the artist ID isn't filled. i guess it is written this way as a failsafe in case MB doesn't have the album, because if it did, the tags would already be filled, but i think maybe it would be better to put this
> 
> https://github.com/rdswift/picard-scripts/blob/d1ee2b13c932f47cc28595c6608db75cafe0bb84/naming_scripts/Bob_Swift/Bob%20Swift's%20Naming%20Script.pts#L241

I didn't do it that way because I everything I tag and file using Picard is based on the entry being in MusicBrainz.  If what I'm tagging isn't in MusicBrainz, I add it before doing the tagging.  In that respect, then yes, it is a failsafe / fallback that I rarely (if ever) end up using in my situation.  If you prefer to use a different fallback for your script that's more in line with your workflow, then by all means use it.

> here where you check if a track has a different artist than the album (this is a great feature btw), ie, if any one track meets this criteria then you could conceivably argue that the album should be categorized as Various Artists?

The issue with that is that Picard processes each track separately (running the script once for each track and saving that track's file before moving to the next track), so there's no way (short of using the functionality of the Persistent Variables plugin in a tagging script) to trigger something for the whole album based on the contents of a single track's metadata.

> either way in my experience albums that aren't on MB get sorted as Various Artists, and in tandem with the soundtracks issue i described in https://github.com/rdswift/picard-scripts/pull/5 , my non-MB soundtracks will get thrown into Various Artists and will have the solo artist of the album listed in the filename in [Brackets] as if they were different from the main album artist

It sounds like your workflow and use case is sufficiently different from mine that you should be making changes to your script to accommodate the differences.

**Comment by ofsaleem** [comment]
thank you for the explanation and breakdown of your philosophy, i understand it a lot better. i think i will take your advice and modify my script so it works better for my needs. thanks for your time! again the script has been a total lifesaver

cheers!
